packages/perpetual/src/fundingRateProposer.js

```diff
@@ -0,0 +1,223 @@
+    this.account = account;
+    this.perpetualFactoryClient = perpetualFactoryClient;
+    this.web3 = this.perpetualFactoryClient.web3;
+    this.commonPriceFeedConfig = commonPriceFeedConfig;
+
+    this.createPerpetualContract = perpAddress => {
+      return new this.web3.eth.Contract(getAbi("Perpetual"), perpAddress);
+    };
+
+    // Gas Estimator to calculate the current Fast gas rate.
+    this.gasEstimator = gasEstimator;
+
+    // Cached mapping of identifiers to pricefeed classes
+    this.priceFeedCache = {};
+
+    // Cached perpetual contracts
+    this.contractCache = {};
+
+    // Helper functions from web3.
+    this.BN = this.web3.utils.BN;
+    this.toBN = this.web3.utils.toBN;
+    this.toWei = this.web3.utils.toWei;
+    this.fromWei = this.web3.utils.fromWei;
+    this.utf8ToHex = this.web3.utils.utf8ToHex;
+    this.hexToUtf8 = this.web3.utils.hexToUtf8;
+
+    // Default config settings. Bot deployer can override these settings by passing in new
+    // values via the `perpetualProposerConfig` input object. The `isValid` property is a function that should be called
+    // before resetting any config settings. `isValid` must return a Boolean.
+    const defaultConfig = {
+      fundingRateErrorPercent: {
+        //   "fundingRateErrorPercent":0.05 ->  Current funding rates (as stored in the OptimisticOracle)
+        //                                      that do not equal the bot's queried funding rate
+        //                                      within this error % will be requested to be updated,
+        //                                      and proposed to.
+        //                                      e.g. 0.05 implies 5% margin of error.
+        value: 0.05,
+        isValid: x => {
+          return x >= 0 && x < 1;
+        }
+      }
+    };
+
+    // Validate and set config settings to class state.
+    const configWithDefaults = createObjectFromDefaultProps(perpetualProposerConfig, defaultConfig);
+    Object.assign(this, configWithDefaults);
+  }
+
+  async update() {
+    await Promise.all([this.perpetualFactoryClient.update(), this.gasEstimator.update()]);
+
+    // Create contract object for each perpetual address created.
+    for (let creationEvent of this.perpetualFactoryClient.getAllCreatedContractEvents()) {
+      if (!this.contractCache[creationEvent.contractAddress]) {
+        // Failure to construct a Perpetual instance using the contract address should be fatal,
+        // so we don't catch that error.
+        const perpetualContract = this.createPerpetualContract(creationEvent.contractAddress);
+        this.contractCache[creationEvent.contractAddress] = perpetualContract;
+      }
+    }
+
+    // Increase allowances for all relevant collateral currencies.
+    await this._setAllowances();
+  }
+
+  async updateFundingRates() {
+    this.logger.debug({
+      at: "PerpetualProposer#updateFundingRates",
+      message: "Checking for contract funding rates to update"
```

Comment:
do you want to add:
```
perpetualsChecked: Object.keys(this.contractCache)
```
to the log to the addresses? might help a bit with debugging.